monash-human-power/MHP_DAS
# Implement Smoothing Techniques on DAS Data

**Updated as of 14/06, read for new screenshots and steps to run**
## Description
The following changes implements 2 smoothing techniques into `das_data_sorter.py` to filter out unwanted data and fluctuations. The smoothing techniques are called 'Moving Mean' and 'Moving Median'. The output will be used for BOOST's Power Model. 

Will write documentation for how Smoothing Technique works after exam period, or after this PR. However, if you want to know how it works now:
- [Video explaining Smoothing Time Series Data](https://www.youtube.com/watch?v=U0dF8uJs3mE)
- [Timestamp for Moving Mean for Odd values of N](https://youtu.be/U0dF8uJs3mE?t=52)
- [Timestamp for Moving Mean for Even values of N](https://youtu.be/U0dF8uJs3mE?t=396)
- [Timestamp for Moving Median for Odd values of N](https://youtu.be/U0dF8uJs3mE?t=676)
- [Timestamp for Moving Median for Even values of N](https://youtu.be/U0dF8uJs3mE?t=1084)

To sum it up, Moving Mean and Moving Median basically 'smooth' the data points. Starting from the first data point, it will take N consecutive number of data points and find the mean/median between them, then it will move up on data point and take the next N consecutive number of data points. By doing this, this will give the average of the entire data and remove noise. An example of 3-point (odd) Moving Mean is shown below:
![image](https://user-images.githubusercontent.com/58575113/83941622-2cb24a80-a830-11ea-96f7-718ddfe14c09.png)

Having a higher value for N means more smoother data points and less fluctuations, however, this will result in having less data points overall. 

Refer down below for photos of how even-numbered N works.

## Screenshots
Left: Regular filtered data with **no** smoothing
Right: Smoothed and filtered data with smoothing
CSV file used: data_291.csv

Mean smoothing using 23 data points:
![image](https://user-images.githubusercontent.com/58575113/84574410-952d9880-ade9-11ea-9845-b3d90fd28729.png)

Median smoothing using 23 data points:
![image](https://user-images.githubusercontent.com/58575113/84574437-c60dcd80-ade9-11ea-875f-d22b22a2b607.png)

## Steps to Test
For converting from milliseconds to seconds of data **(no smoothing)**:
1. In the terminal, run `python das_data_sorter.py --input <file name> --output <output name> --unit <seconds/minutes>`
Unit has a default of seconds
2. When finished, it should print 'Success!' into the terminal

For applying smoothing techniques:
1. In the terminal, run `python das_data_sorter.py --input <file name> --output <output name> --unit <seconds/minutes> --smooth <mean/median> -n <number of data points>`
2. When finished, it should print 'Success!' into the terminal
